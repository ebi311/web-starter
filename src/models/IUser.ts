export type Gender = 'man' | 'woman' | 'none';
export default interface IUser {
  familyName: string;
  givenName: string;
  gender: Gender;
}
