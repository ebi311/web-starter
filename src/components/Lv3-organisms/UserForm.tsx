import React, { useCallback, useMemo } from 'react';
import { useDispatch } from 'react-redux';
import { setUser } from '../../actions/UserActions';
import IUser, { Gender } from '../../models/IUser';
import LabeledTetBox from '../Lv2-molecules/LabeledTextBox';
import LabeledRadioButton from '../Lv2-molecules/LabeledRadioButton';
import { styled } from '../Theme';

const Container = styled.div``;

const TextBox = styled(LabeledTetBox)`
  margin-bottom: 1em;
`;

const RadioList = styled(LabeledRadioButton)`
  margin-bottom: 1em;
`;

interface IProps {
  user: IUser;
}

const UserForm: React.FC<IProps> = props => {
  const user = useMemo(() => props.user, [props.user]);
  const dispatch = useDispatch();
  const onFamilyNameOnChange = useCallback((value: string) => {
    dispatch(setUser({ familyName: value }));
  }, []);
  const onGivenNameOnChange = useCallback((value: string) => {
    dispatch(setUser({ givenName: value }));
  }, []);
  const onGenderOnChange = useCallback((value: string) => {
    dispatch(setUser({ gender: value as Gender }));
  }, []);
  const genderSelect = useMemo(
    (): { label: string; value: Gender }[] => [
      {
        label: '男',
        value: 'man',
      },
      {
        label: '女',
        value: 'woman',
      },
      {
        label: '回答しない',
        value: 'none',
      },
    ],
    [],
  );
  return (
    <Container>
      <TextBox
        label="姓"
        value={user.familyName}
        onChange={onFamilyNameOnChange}
      />
      <TextBox
        label="名"
        value={user.givenName}
        onChange={onGivenNameOnChange}
      />
      <RadioList
        label="性別"
        name="gender"
        selectItems={genderSelect}
        value={user.gender}
        onChange={onGenderOnChange}
      />
    </Container>
  );
};

export default UserForm;
