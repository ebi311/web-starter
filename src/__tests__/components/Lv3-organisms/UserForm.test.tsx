/* eslint-disable @typescript-eslint/no-explicit-any */
import { shallow } from 'enzyme';
import React from 'react';
import UserForm from '../../../components/Lv3-organisms/UserForm';
import IUser from '../../../models/IUser';

const dispatch = jest.fn();

jest.mock('react-redux', () => ({
  useDispatch: () => dispatch,
}));

jest.mock('../../../actions/UserActions', () => ({
  setUser: jest.fn().mockImplementation((obj: any) => obj),
}));

describe('UserForm', () => {
  const user: IUser = {
    familyName: 'nanashi',
    givenName: 'gonbe',
    gender: 'man',
  };
  test('success', () => {
    const wrapper = shallow(<UserForm user={user} />).dive();
    expect(wrapper).toMatchSnapshot();
    const FamilyNameTextbox = wrapper.find('[label="姓"]');
    FamilyNameTextbox.simulate('change', 'f001');
    expect(dispatch).toBeCalledWith({ familyName: 'f001' });
    dispatch.mockClear();
    const GivenNameTextbox = wrapper.find('[label="名"]');
    GivenNameTextbox.simulate('change', 'g001');
    expect(dispatch).toBeCalledWith({ givenName: 'g001' });
    dispatch.mockClear();
    const GenderRadioButton = wrapper.find('[label="性別"]');
    GenderRadioButton.simulate('change', 'gen001');
    expect(dispatch).toBeCalledWith({ gender: 'gen001' });
  });
});
