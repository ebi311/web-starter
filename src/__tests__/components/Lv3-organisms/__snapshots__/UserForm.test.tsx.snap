// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm success 1`] = `
<div
  className="sc-AxgMl fFPQpl"
>
  <Styled(LabeledTetBox)
    label="姓"
    onChange={[Function]}
    value="nanashi"
  />
  <Styled(LabeledTetBox)
    label="名"
    onChange={[Function]}
    value="gonbe"
  />
  <Styled(LabeledRadioButton)
    label="性別"
    name="gender"
    onChange={[Function]}
    selectItems={
      Array [
        Object {
          "label": "男",
          "value": "man",
        },
        Object {
          "label": "女",
          "value": "woman",
        },
        Object {
          "label": "回答しない",
          "value": "none",
        },
      ]
    }
    value="man"
  />
</div>
`;
